engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/schemaregistry/confluent/formatter/JsonPayloadToJsonFormatterFactory.scala

```diff
@@ -0,0 +1,63 @@
+package pl.touk.nussknacker.engine.avro.schemaregistry.confluent.formatter
+
+import io.circe.Json
+import org.apache.flink.api.common.typeinfo.TypeInformation
+import org.apache.flink.streaming.connectors.kafka.KafkaDeserializationSchema
+import org.apache.kafka.clients.consumer.ConsumerRecord
+import org.apache.kafka.common.serialization.Deserializer
+import pl.touk.nussknacker.engine.api.CirceUtil
+import pl.touk.nussknacker.engine.avro.RuntimeSchemaData
+import pl.touk.nussknacker.engine.avro.serialization.KafkaAvroKeyValueDeserializationSchemaFactory
+import pl.touk.nussknacker.engine.kafka.consumerrecord.ConsumerRecordToJsonFormatter
+import pl.touk.nussknacker.engine.kafka.{KafkaConfig, RecordFormatter, RecordFormatterFactory}
+
+import java.nio.charset.StandardCharsets
+import scala.reflect.ClassTag
+
+/**
+  * RecordFormatter factory for kafka avro sources with json payload.
+  *
+  * Test data record is a simple record representing data of ConsumerRecord. It does not contain schema ids.
+  * Creates instance of ConsumerRecordToJsonFormatter with fixed key-value types (key and value are always deserialized to Json).
+  */
+class JsonPayloadToJsonFormatterFactory extends RecordFormatterFactory {
+
+  override def create[K: ClassTag, V: ClassTag](kafkaConfig: KafkaConfig, kafkaSourceDeserializationSchema: KafkaDeserializationSchema[ConsumerRecord[K, V]]): RecordFormatter = {
+    val asJsonDeserializerFactory = new KeyValueKafkaJsonDeserializerFactory
+    val asJsonDeserializer = asJsonDeserializerFactory.create[Json, Json](kafkaConfig, None, None)
+
+    new ConsumerRecordToJsonFormatter[Json, Json](asJsonDeserializer)
+  }
+
+}
+
+class KeyValueKafkaJsonDeserializerFactory extends KafkaAvroKeyValueDeserializationSchemaFactory {
+
+  override protected def createKeyDeserializer[K: ClassTag](schemaDataOpt: Option[RuntimeSchemaData], kafkaConfig: KafkaConfig): Deserializer[K] =
+    toJsonDeserializer.asInstanceOf[Deserializer[K]]
+
+  override protected def createKeyTypeInfo[K: ClassTag](schemaDataOpt: Option[RuntimeSchemaData], kafkaConfig: KafkaConfig): TypeInformation[K] =
+    TypeInformation.of(classOf[Json]).asInstanceOf[TypeInformation[K]]
+
+  override protected def createValueDeserializer[V: ClassTag](schemaDataOpt: Option[RuntimeSchemaData], kafkaConfig: KafkaConfig): Deserializer[V] =
+    toJsonDeserializer.asInstanceOf[Deserializer[V]]
+
+  override protected def createValueTypeInfo[V: ClassTag](schemaDataOpt: Option[RuntimeSchemaData], kafkaConfig: KafkaConfig): TypeInformation[V] =
+    TypeInformation.of(classOf[Json]).asInstanceOf[TypeInformation[V]]
+
+  // always deserialize key to valid Json
+  override protected def createKeyOrStringDeserializer[K: ClassTag](schemaDataOpt: Option[RuntimeSchemaData], kafkaConfig: KafkaConfig): Deserializer[K] =
+    if (kafkaConfig.useStringForKey) {
```

Comment:
Won't be easier to instead of extracting method `createKeyOrStringDeserializer` extract method `createStringKeyDeserializer: Deserializer[_]`? You would remove duplication of this if-else then.

Reply:
True, it is easier that way :) Fixed.